Home connect functional and ambient light added

Sjack-Sch opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


  This PR adds functional and ambient light for home connect devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
Usage
INSTALLATION USING A DEVELOPER ACCOUNT

Visit https://developer.home-connect.com and sign up for a developer account.
Under Applications, register a new App:
Application ID: Home Assistant (or whatever name makes sense to you)
OAuth Flow: Authorization Code Grant Flow
Redirect URI: “<INTERNAL_HOME_ASSISTANT_URL>/auth/external/callback Use your internal Home Assistant URL, if you didn’t configure one manually, use your local IP address. Examples: http://192.168.0.2:8123/auth/external/callback, http://homeassistant.local:8123/auth/external/callback.”
Add the client ID and client secret to your configuration.yaml (see below).
Once Home Assistant has restarted, go to Configuration > Integrations.
Select the Home Connect integration.
You will be prompted to open a web page and authenticate your app with Home Connect.
Use your Home Connect account details associated with your developer app (this may not be your developer account details you created but will be the account you use to login to the original Home Connect App from Bosch/Siemens).
You will know it works when you are prompted to allow Home Assistant to access your Home Connect data, including the list of devices accessible to the original Home Connect app.
-->

```yaml
# Example configuration.yaml
# Example configuration.yaml entry

home_connect:
  client_id: CLIENT_ID
  client_secret: CLIENT_SECRET
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15960

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Sjack-Sch committed:
Update nl.json

Sjack-Sch committed:
added period to end of every logging entry.

Sjack-Sch committed:
Functional and ambient light added

Sjack-Sch committed:
Change to dict.get method

Sjack-Sch committed:
Update light.py

homeassistant commented:
Hi @Sjack-Sch, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44091).

Once you do that we will be able to review and accept this pull request.

Thanks!

rwunsch commented:
This pull-request would be SUPER to have in the NABU CASA "Home Connect"-Integration!

see: https://community.home-assistant.io/t/home-connect-integration-control-exhaust-hood-light/208064/14?u=wunsch

Sjack-Sch commented:
> This pull-request would be SUPER to have in the NABU CASA "Home Connect"-Integration!
> 
> see: https://community.home-assistant.io/t/home-connect-integration-control-exhaust-hood-light/208064/14?u=wunsch
> 
> Thanks, @Sjack-Sch, for adding this feature!
> 
> Could you sign the CLA as described above?
> 
> Thanks and cheers - Robert

Hi Robert,

CLA is signed.

rwunsch commented:
Thanks a ton, @Sjack-Sch !
Let's see if this pull-request now makes it into the solution.
It seems that "all checks have passed" :)

allmoney commented:
In which release version of HA it will be added?

MartinHjelmare commented on homeassistant/components/home_connect/__init__.py:
```diff
@@ -32,7 +32,7 @@
 )
 
-PLATFORMS = ["binary_sensor", "sensor", "switch"]
+PLATFORMS = ["binary_sensor", "sensor", "switch", "light"]
```
Please sort 🔡.

Sjack-Sch replied:
Platforms sorted

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -56,6 +56,8 @@ def get_devices(self):
             elif app.type == "Washer":
                 device = Washer(self.hass, app)
+            elif app.type == "WasherDryer":
+                device = WasherDryer(self.hass, app)
```
This doesn't seem related to the light platform. Please don't mix different changes in the same PR. Make multiple PRs each focusing on one thing.

Sjack-Sch replied:
Removed all none light platform related changes.

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -69,7 +71,7 @@ def get_devices(self):
                 device = Hob(self.hass, app)
             else:
-                _LOGGER.warning("Appliance type %s not implemented", app.type)
+                _LOGGER.warning("Appliance type %s not implemented.", app.type)
```
We don't end logging messages with period.

Sjack-Sch replied:
Period removed.

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -87,21 +89,22 @@ def __init__(self, hass, appliance):
         """Initialize the device class."""
         self.hass = hass
         self.appliance = appliance
+        self.entities = []
```
We shouldn't store entities.

Sjack-Sch replied:
Storing entities removed.

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -324,6 +415,21 @@ class CoffeeMaker(DeviceWithPrograms):
         {"name": "ConsumerProducts.CoffeeMaker.Program.Beverage.WarmMilk"},
         {"name": "ConsumerProducts.CoffeeMaker.Program.Beverage.Ristretto"},
         {"name": "ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.Cortado"},
+        {"name": "ConsumerProducts.CoffeeMaker.Program.CoffeeWorld.CafeCortado"},
```
Please make a separate PR for these changes.

Sjack-Sch replied:
Removed.

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -332,10 +438,13 @@ def get_entity_info(self):
         program_switches = self.get_program_switches()
-        return {"switch": program_switches, "sensor": program_sensors}
+        return {
+            "switch": program_switches,
```
Not needed formatting change.

Sjack-Sch replied:
Formating reverted.

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,209 @@
+        self._hs_color = None
+        self._ambient = ambient
+        if self._ambient:
+            self._brightnesskey = BSH_AMBIENTLIGHTBRIGHTNESS
```
Please separate words with underscore.

Sjack-Sch replied:
All words seperated with underscore.

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,209 @@
+                            await self.hass.async_add_executor_job(
+                                self.device.appliance.set_setting,
+                                self._customcolorkey,
+                                "#" + hex_val,
```
Please use f-strings instead of string concatenation.

Sjack-Sch replied:
Done

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,209 @@
+                            _LOGGER.error(
+                                "Error while trying setting the color: %s", err
+                            )
+                            self._state = False
```
Why are we setting state to False on error?

Sjack-Sch replied:
Removed self._state = false when color setting did not succeed.

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,209 @@
+                            )
+                            self._state = False
+            else:
+                _LOGGER.debug("Tried to switch light on for: %s", self.name)
```
This sounds like we tried and something failed.

Sjack-Sch replied:
For consistency reason I used the same logging as done in the existing platform switch.
See switch.py line 115:
`    async def async_turn_off(self, **kwargs):
        """Switch the device off."""
        _LOGGER.debug("tried to switch off %s", self.name)
        try:
            await self.hass.async_add_executor_job(
                self.device.appliance.set_setting,
                BSH_POWER_STATE,
                self.device.power_off_state,
            )
        except HomeConnectError as err:
            _LOGGER.error("Error while trying to turn off device: %s", err)
            self._state = True
        self.async_entity_update()
`

MartinHjelmare replied:
We should say that we're going to do something if we log before doing that thing. Eg "Switching on bla bla". We can clean up the existing logs in a clean up PR later.

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,209 @@
+
+    async def async_turn_off(self, **kwargs):
+        """Switch the light off."""
+        _LOGGER.debug("tried to switch light off for: %s", self.name)
```
Please start logging messages with capital letter.

Sjack-Sch replied:
Done

MartinHjelmare commented on homeassistant/components/home_connect/translations/nl.json:
```diff
@@ -1,7 +1,8 @@
         "abort": {
-            "missing_configuration": "Het component is niet geconfigureerd. Volg de documentatie."
+            "missing_configuration": "Het component is niet geconfigureerd. Volg de documentatie.",
+            "no_url_available": "Geen URL beschikbaar. Voor meer informatie over de fout, [raadpleeg de help sectie]({docs_url})"
```
Please don't change translation files besides the default English file.

Sjack-Sch replied:
Reverted change on translation file.

Sjack-Sch committed:
Update __init__.py

Platforms sorted 🔡.

Sjack-Sch committed:
Update api.py

- Removed all none light platform related changes.
- Period removed from loggin.
- Storing entities removed.
- Not needed formating change reverted.
-

Sjack-Sch committed:
Update const.py

All words seperated with underscore.

Sjack-Sch committed:
Update nl.json

Reverted change on translation file.

Sjack-Sch committed:
Update light.py

-All words of constants seperated with underscore.
- f-string used iso concatenation.
- Added "ambient"to loggin text.
- Removed self._state = false when color setting did not succeed.
- Logging starts with a capital.

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -93,15 +93,15 @@ def initialize(self):
             self.appliance.get_status()
         except (HomeConnectError, ValueError):
-            _LOGGER.debug("Unable to fetch appliance status. Probably offline")
+            _LOGGER.debug("Unable to fetch appliance status. Probably offline.")
```
There are still changes that aren't addressed from my comments.

Sjack-Sch replied:
Sorry, some have been forgotten.

MartinHjelmare commented on homeassistant/components/home_connect/api.py:
```diff
@@ -369,4 +399,7 @@ def get_entity_info(self):
+        return {
+            "switch": program_switches,
+            "sensor": program_sensors,
+        }
```
Revert this change.

Sjack-Sch replied:
Done.

Sjack-Sch replied:
Black is complaining on:   return {"switch": program_switches,"sensor": program_sensors}

MartinHjelmare commented on homeassistant/components/home_connect/entity.py:
```diff
@@ -20,6 +20,7 @@ def __init__(self, device: HomeConnectDevice, desc: str) -> None:
         self.device = device
         self.desc = desc
         self._name = f"{self.device.appliance.name} {desc}"
+        self.device.entities.append(self)
```
Remove.

Sjack-Sch replied:
Done

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+        if self._ambient:
+            if ATTR_BRIGHTNESS in kwargs or ATTR_HS_COLOR in kwargs:
+                try:
+                    await self.hass.async_add_executor_job(
```
Maybe this is a comment for another PR, but it seems wasteful to use our async api when the library interface needs to run in sync context.

We should probably adjust the entity parent class in a followup  PR so that we can update state with the sync api `self.schedule_update_ha_state`.

Sjack-Sch replied:
OK

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                )
+            except HomeConnectError as err:
+                _LOGGER.error("Error while trying to turn on light: %s", err)
+                self._state = False
```
We shouldn't change state on error. Possibly set the entity to unavailable if that's what this error means.

Sjack-Sch replied:
For consistency reason I followed the same solution as done in the platform switch. 
See switch.py L101

`    async def async_turn_on(self, **kwargs):
        """Switch the device on."""
        _LOGGER.debug("Tried to switch on %s", self.name)
        try:
            await self.hass.async_add_executor_job(
                self.device.appliance.set_setting,
                BSH_POWER_STATE,
                BSH_POWER_ON,
            )
        except HomeConnectError as err:
            _LOGGER.error("Error while trying to turn on device: %s", err)
            self._state = False
        self.async_entity_update()`

MartinHjelmare replied:
We shouldn't do that. We don't know that that's the correct thing to do.

MartinHjelmare replied:
Bump

Sjack-Sch replied:
removed state setting after error

rwunsch commented:
Thank you both, @Sjack-Sch and  @MartinHjelmare !!
I really appreciate your hard work and diligent effort.
I am inspired, and hope I will pick-up Python and the structure of home-assistant sometime soon, to be able to contribute!
THANKS A TON!

Sjack-Sch committed:
Update api.py

- Removed ending perio in logging
- Reverted formating
- Removed self.device.entities.append(self)
-

Sjack-Sch committed:
Update entity.py

- Removed self.device.entities.append(self)

Sjack-Sch committed:
Update api.py

- Adding newline at end of file
- Added whitespave after ","

Sjack-Sch committed:
Update api.py

Newline at end of file

MartinHjelmare commented on homeassistant/components/home_connect/const.py:
```diff
@@ -11,6 +11,20 @@
+BSH_AMBIENT_LIGHT_CUSTOM_COLOR = "BSH.Common.Setting.AmbientLightCustomColor"
+
 BSH_DOOR_STATE = "BSH.Common.Status.DoorState"
+BSH_PAUSE = "BSH.Common.Command.PauseProgram"
+BSH_RESUME = "BSH.Common.Command.ResumeProgram"
```
These aren't used.
```suggestion
```

Sjack-Sch replied:
Removed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+        self._hs_color = None
+        self._ambient = ambient
+        if self._ambient:
+            self._brightnesskey = BSH_AMBIENT_LIGHT_BRIGHTNESS
+            self._key = BSH_AMBIENT_LIGHT_ENABLED
+            self._customcolorkey = BSH_AMBIENT_LIGHT_CUSTOM_COLOR
+            self._colorkey = BSH_AMBIENT_LIGHT_COLOR
+        else:
+            self._brightnesskey = COOKING_LIGHTING_BRIGHTNESS
+            self._key = COOKING_LIGHTING
+            self._customcolorkey = None
+            self._colorkey = None
```
```suggestion
            self._brightness_key = BSH_AMBIENT_LIGHT_BRIGHTNESS
            self._key = BSH_AMBIENT_LIGHT_ENABLED
            self._custom_color_key = BSH_AMBIENT_LIGHT_CUSTOM_COLOR
            self._color_key = BSH_AMBIENT_LIGHT_COLOR
        else:
            self._brightness_key = COOKING_LIGHTING_BRIGHTNESS
            self._key = COOKING_LIGHTING
            self._custom_color_key = None
            self._color_key = None
```

Sjack-Sch replied:
Oeps, changed.

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+            try:
+                await self.hass.async_add_executor_job(
+                    self.device.appliance.set_setting,
+                    self._brightnesskey,
```
```suggestion
                    self._brightness_key,
```

Sjack-Sch replied:
Changed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                _LOGGER.debug("Updated, new brightness: %s", self._brightness)
+
+        else:
+            brightness = self.device.appliance.status.get(self._brightnesskey, {})
```
```suggestion
            brightness = self.device.appliance.status.get(self._brightness_key, {})
```

Sjack-Sch replied:
Changed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                        try:
+                            await self.hass.async_add_executor_job(
+                                self.device.appliance.set_setting,
+                                self._customcolorkey,
```
```suggestion
                                self._custom_color_key,
```

Sjack-Sch replied:
Changed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+        _LOGGER.debug("Updated, new light state: %s", self._state)
+
+        if self._ambient:
+            color = self.device.appliance.status.get(self._customcolorkey, {})
```
```suggestion
            color = self.device.appliance.status.get(self._custom_color_key, {})
```

Sjack-Sch replied:
Changed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                try:
+                    await self.hass.async_add_executor_job(
+                        self.device.appliance.set_setting,
+                        self._colorkey,
```
```suggestion
                        self._color_key,
```

Sjack-Sch replied:
Changed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                                "Error while trying setting the color: %s", err
+                            )
+            else:
+                _LOGGER.debug("Tried to switch ambient light on for: %s", self.name)
```
```suggestion
                _LOGGER.debug("Trying to switch ambient light on for: %s", self.name)
```

Sjack-Sch replied:
Changed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                    self._state = False
+
+        elif ATTR_BRIGHTNESS in kwargs:
+            _LOGGER.debug("Tried to change brightness for: %s", self.name)
```
```suggestion
            _LOGGER.debug("Changing brightness for: %s", self.name)
```

Sjack-Sch replied:
Changed

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                _LOGGER.error("Error while trying set the brightness: %s", err)
+                self._brightness = None
+        else:
+            _LOGGER.debug("Tried to switch light on for: %s", self.name)
```
```suggestion
            _LOGGER.debug("Switching light on for: %s", self.name)
```

Sjack-Sch replied:
Updated

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+
+    async def async_turn_off(self, **kwargs):
+        """Switch the light off."""
+        _LOGGER.debug("Tried to switch light off for: %s", self.name)
```
```suggestion
        _LOGGER.debug("Switching light off for: %s", self.name)
```

Sjack-Sch replied:
Updated

Sjack-Sch committed:
Update const.py

Removed unused.

Sjack-Sch committed:
Update light.py

- seperated words with whitespaces
- improved debug text

MartinHjelmare commented:
If you would have committed my suggestions in the github UI the comments would have been resolved automatically and I wouldn't have to look at those changes again.

Sjack-Sch commented:
> If you would have committed my suggestions in the github UI the comments would have been resolved automatically and I wouldn't have to look at those changes again.

Hi Martin,
I need some guides here. I use GitHub Desktop and Visual Studio Code on macOS but I cannot find a way to resolve comments in either UIs.

MartinHjelmare commented on homeassistant/components/home_connect/light.py:
```diff
@@ -0,0 +1,207 @@
+                    _LOGGER.error(
+                        "Error while trying to turn on ambient light: %s", err
+                    )
+                    self._state = False
```
We're still changing state on error. I don't think we should do that. We can't guarantee that this is the correct current state.

Sjack-Sch replied:
OK, I will remove the state setting after an error.

@DavidMStraub, for consistency reason we should probably create a PR to do the same for switch platform.

Sjack-Sch committed:
Update light.py

remove state setting after an error

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
Please link a docs PR in the PR description. We need to update the docs `ha_category` with the new platform category.

Sjack-Sch commented:
> Please link a docs PR in the PR description. We need to update the docs `ha_category` with the new platform category.

Doc PR [https://github.com/home-assistant/home-assistant.io/pull/15960](url) added.
